compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -620,6 +620,55 @@ impl<'a> Parser<'a> {
             );
         }
 
+        // Possible missing `struct` or `enum` keyword
+        if let TokenKind::Ident(_, _) = self.prev_token.kind
+            && let TokenKind::OpenDelim(Delimiter::Brace) = self.token.kind
+            && let [TokenType::Token(token::Not), TokenType::Token(token::PathSep)] = expected[..]
+        {
+            let mut lookahead = self.clone();
+            for position in 1..1000 {
```

Comment:
This arbitrary lookahead is also unnecessary. For structs, you need to look for a colon, for enums, you need to look for an identifier not followed by a colon. (either unit enum `enum X { A, B, C }` or with data `enum Y { D(E), F(G) }`, `enum Z { H { i: J } }`)

All of these happen under two tokens after the opening brace.

If you find a close brace before finding any of these characteristics, then a `struct` and a `enum` are both possible, so probably suggest both